### PR TITLE
record panic and optionally stack traces

### LINKED ISSUE #12: panic doesn't mark span as error
the middleware doesn't call recover anywhere, pressumably because otel already does that according to the trace

```
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
	/home/aep/.go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.11.0/trace/span.go:396 +0x2a
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc00033e480, {0x0, 0x0, 0xc0000b4780?})
	/home/aep/.go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.11.0/trace/span.go:435 +0x8ee
```

this records an event, but does not set the span as error.

confusingly, a 404 is set as error,  ~~but i don't understand where that's happening either.~~   because otelchi calls span.SetStatus . should it also do that for panic, or is that a problem in otel?

### PR DESCRIPTION
**Pull request by aep**
fixes #12

### CODE REVIEW

**Review by riandyrn — CHANGES_REQUESTED** [review]
Hello, @aep

Thanks a lot for noticing issue #12 & created this PR!

However can you also add test to [middleware_test.go](https://github.com/riandyrn/otelchi/blob/master/middleware_test.go)? This is to ensure otelchi will always work as expected in future releases.

Thanks

File: middleware.go [hunk 1]
```diff
@@ -139,7 +145,30 @@ func (tw traceware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		oteltrace.WithAttributes(semconv.HTTPServerAttributesFromHTTPRequest(tw.serverName, routePattern, r)...),
 		oteltrace.WithSpanKind(oteltrace.SpanKindServer),
 	)
-	defer span.End()
+	defer func() {
+		if recovered := recover(); recovered != nil {
+			defer panic(recovered)
```
Comment (knightazura) [inline middleware.go:150]: May I know, why `panic` called again in here? Is it intended?
At first, I thought this PR will handle if `panic` happens it won't crash the app.

File: middleware.go [hunk 2]
```diff
@@ -139,7 +145,30 @@ func (tw traceware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+					semconv.ExceptionTypeKey.String(typeStr(recovered)),
+					semconv.ExceptionMessageKey.String(fmt.Sprint(recovered)),
+				),
+			}
+
+			if tw.recordStackTrace {
+				opts = append(opts, oteltrace.WithAttributes(
+					semconv.ExceptionStacktraceKey.String(recordStackTrace()),
+				))
+			}
```
Comment (knightazura) [inline middleware.go:165]: Have you check if this attribute recorded properly?
I've create a test to check this, but those attributes not found in the `span`.

You can see the test in here
https://github.com/knightazura/otelchi/blob/59d949aa1b25b6b3230f7ac09e352f11c2a50e46/middleware_test.go#L378-L437

**Comment by riandyrn** [comment]
Hello, @aep

Any update for this PR?

File: middleware.go (hunk 1 above)
Comment (kedare) [reply middleware.go:150]: Hmm for me it should rethrow the panic, its job is just to record it, not to stop it.

**Comment by riandyrn** [comment]
Hello, @aep

After some thorough checking, I don't think this PR is necessary.

There are 2 reasons for this:

1. When the panic occurs, OpenTelemetry does record the error (see [here](https://github.com/open-telemetry/opentelemetry-go/blob/84e28fd3bdb10b11a15be895f2eda56f2c341384/sdk/trace/span.go#L381-L398) for details).
2. The panic handling should be done from end-user side, because maybe the panic is expected and the user doesn't want to treat that as error.

Based on these reasons I'm closing this PR & #12.

Thank you so much for your contribution, @aep!

**Note:**

If you think my decision is incorrect, feel free to re-open this. Thanks.

**Pull request closed without merging**